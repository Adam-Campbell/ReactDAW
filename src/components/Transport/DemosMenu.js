import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import { preludeData, divinityData, chilledData } from '../../demoData';

const DemosMenu = (props) => (
    <div 
        className="menu__container" 
        ref={props.containerRef}
        onClick={props.toggleMenu}
    >
        <p 
            className={`menu__label ${props.isOpen ? 'menu__label--open' : ''}`}
        >Demos</p>
        <ul 
            className={`menu__list ${props.isOpen ? 'menu__list--visible' : ''}`}
        >
            <li 
                className="menu__item"
                onClick={() => props.loadDemo('Prelude_demo', preludeData)}
            >
                <p className="menu__item-text">Prelude</p>
            </li>
            <li 
                className="menu__item"
                onClick={() => props.loadDemo('Divinity_demo', divinityData)}
            >
                <p className="menu__item-text">Divinity</p>
            </li>
            <li 
                className="menu__item"
                onClick={() => props.loadDemo('Chilled_demo', chilledData)}
            >
                <p className="menu__item-text">Chilled</p>
            </li>
        </ul>
    </div>
);

DemosMenu.propTypes = {
    isOpen: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
    containerRef: PropTypes.object.isRequired
};

export default connect(
    undefined,
    {
        loadDemo: ActionCreators.loadStateSuccess
    }
)(DemosMenu);