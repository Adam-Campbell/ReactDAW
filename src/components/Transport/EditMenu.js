import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import { ActionCreators as undoableActions } from 'redux-undo'; 

export const EditMenu = props => {
    return (
        <div 
            className="menu__container" 
            ref={props.containerRef}
            onClick={props.toggleMenu}
        >
            <p 
                className="menu__label"
            >Edit</p>
            <ul 
                className={`menu__list ${props.isOpen ? 'menu__list--visible' : ''}`}
            >
                <li 
                    className="menu__item"
                    onClick={props.undo}
                >
                    <p className="menu__item-text">Undo</p>
                </li>
                <li 
                    className="menu__item"
                    onClick={props.redo}
                >
                    <p className="menu__item-text">Redo</p>
                </li>
            </ul>
        </div>
    );
}

EditMenu.propTypes = {
    isOpen: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
    containerRef: PropTypes.object.isRequired
}

export default connect(
    undefined, 
    {
        undo: undoableActions.undo,
        redo: undoableActions.redo
    }
)(EditMenu);