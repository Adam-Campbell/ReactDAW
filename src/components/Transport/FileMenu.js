import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import { modalTypes } from '../../constants';

export const FileMenu = props => {

    const handleSave = () => {
        if (props.saveName) {
            const namespacedSaveName = `[track_name] ${props.saveName}`;
            props.saveState(namespacedSaveName);
        } else {
            props.openModal(modalTypes.saveAs);
        }
    }

    return (
        <div 
            className="menu__container" 
            ref={props.containerRef}
            onClick={props.toggleMenu}
        >
            <p 
                className="menu__label"
            >File</p>
            <ul 
                className={`menu__list ${props.isOpen ? 'menu__list--visible' : ''}`}
            >
                <li 
                    className="menu__item"
                    onClick={() => props.openModal(modalTypes.open)}
                >
                    <p className="menu__item-text">Open</p>
                </li>
                <li 
                    className="menu__item"
                    onClick={handleSave}
                >
                    <p className="menu__item-text">Save</p>
                </li>
                <li 
                    className="menu__item"
                    onClick={() => props.openModal(modalTypes.saveAs)}
                >
                    <p className="menu__item-text">Save As</p>
                </li>
            </ul>
        </div>
    );
}

FileMenu.propTypes = {
    isOpen: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
    containerRef: PropTypes.object.isRequired
}

const mapStateToProps =  state => ({
    saveName: state.main.present.composition.saveName
});

export default connect(
    mapStateToProps,
    { 
        openModal: ActionCreators.openModal,
        saveState: ActionCreators.saveState
    }
)(FileMenu);