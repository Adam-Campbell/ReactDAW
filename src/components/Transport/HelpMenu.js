import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import { modalTypes } from '../../constants';

export const HelpMenu = props => {
    return (
        <div 
            className="menu__container" 
            ref={props.containerRef}
            onClick={props.toggleMenu}
        >
            <p 
                className={`menu__label ${props.isOpen ? 'menu__label--open' : ''}`}
            >Help</p>
            <ul 
                className={`menu__list ${props.isOpen ? 'menu__list--visible' : ''}`}
            >
                <li 
                    className="menu__item"
                    onClick={() => props.openModal(modalTypes.shortcuts)}
                >
                    <p className="menu__item-text">View shortcuts</p>
                </li>
            </ul>
        </div>
    );
}

HelpMenu.propTypes = {
    isOpen: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
    containerRef: PropTypes.object.isRequired
}

export default connect(
    undefined,
    {
        openModal: ActionCreators.openModal
    }
)(HelpMenu);