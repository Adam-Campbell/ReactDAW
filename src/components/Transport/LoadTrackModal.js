import React from 'react';
import PropTypes from 'prop-types';

const LoadTrackModal = props => (
    <div className="modal__overlay">
        <div className="modal__content-container">
            <button
                className="button pink"
                onClick={props.handleClose}
            >Close</button>
            <ul className="modal__list">
            {props.savedTracks.map((trackName, index) => (
                <li
                    className="modal__list-item"
                    key={index}
                    onClick={(e) => props.handleLoad(trackName)}
                >{trackName}</li>
            ))}
            </ul>
        </div>
    </div>
);

LoadTrackModal.propTypes = {
    handleClose: PropTypes.func.isRequired,
    savedTracks: PropTypes.arrayOf(PropTypes.string).isRequired,
    handleLoad: PropTypes.func.isRequired
};

export default LoadTrackModal;