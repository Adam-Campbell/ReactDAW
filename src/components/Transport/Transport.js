import React from 'react';
import PropTypes from 'prop-types';

const Transport = props => (
    <div 
        className="transport__container"
        onClick={props.handleTransportBarClick}
    >
        <button 
            className="button pink"
            onClick={props.playTrack}
        >Play</button>
        <button 
            className="button pink"
            onClick={props.stopTrack}
        >Stop</button>
        <span className="transport__track-position" >{props.trackPosition}</span>
        <div className="transport__bpm-container">
            <span className="transport__bpm-label">BPM:</span>
            {props.isEditingBPM ? 
                <input
                    className="transport__bpm-input"
                    value={props.editedBPM}
                    onChange={props.handleBPMChange}
                    ref={props.inputRef}
                    onClick={e => e.stopPropagation()}
                ></input> :
                <p
                    className="transport__bpm-text"
                    onClick={props.enterBPMEditingMode}
                >{props.bpm}</p>
            }
        </div>
    </div>
);

Transport.propTypes = {
    handleTransportBarClick: PropTypes.func.isRequired,
    playTrack: PropTypes.func.isRequired,
    stopTrack: PropTypes.func.isRequired,
    trackPosition: PropTypes.string.isRequired,
    isEditingBPM: PropTypes.bool.isRequired,
    editedBPM: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    handleBPMChange: PropTypes.func.isRequired,
    inputRef: PropTypes.object.isRequired,
    enterBPMEditingMode: PropTypes.func.isRequired,
    bpm: PropTypes.number.isRequired
};

export default Transport;