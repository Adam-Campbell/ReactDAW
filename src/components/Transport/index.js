import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import Tone from 'tone';
import Transport from './Transport';
import { formatTransportPosition } from './transportUtils';

export class TransportContainer extends Component {
    constructor(props) {
        super(props);
        this.rAFRef = null;
        this.inputRef = React.createRef();
        this.state = {
            transportPosition: Tone.Transport.position,
            isEditingBPM: false,
            editedBPM: this.props.bpm
        }
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.isPlaying !== this.props.isPlaying) {
            if (this.props.isPlaying) {
                requestAnimationFrame(this.getTransportPosition);
            } else {
                cancelAnimationFrame(this.rAFRef);
                this.setState({
                    transportPosition: '0:0:0'
                });
            }
        }
    }

    getTransportPosition = () => {
        const newTransportPosition = formatTransportPosition(Tone.Transport.position);
        if (newTransportPosition !== this.state.transportPosition) {
            this.setState({
                transportPosition: newTransportPosition
            });
        }
        this.rAFRef = requestAnimationFrame(this.getTransportPosition);
    }

    handleBPMChange = (e) => {
        this.setState({
            editedBPM: e.target.value
        });
    }

    enterBPMEditingMode = () => {
        this.setState({
            isEditingBPM: true
        }, () => {
            this.inputRef.current.focus();
        });
    }

    exitBPMEditingMode = () => {
        const bpmString = this.state.editedBPM;
        const bpmNum = parseInt(bpmString);
        if (!Number.isNaN(bpmNum)) {
            this.props.setBPM(bpmNum);
            this.setState({
                isEditingBPM: false
            });
        } else {
            this.setState({
                isEditingBPM: false,
                editedBPM: this.props.bpm
            });
        }
    }

    handleTransportBarClick = () => {
        if (this.state.isEditingBPM) {
            this.exitBPMEditingMode();
        }
    }

    render() {
        return <Transport 
            handleTransportBarClick={this.handleTransportBarClick}
            playTrack={this.props.playTrack}
            stopTrack={this.props.stopTrack}
            transportPosition={this.state.transportPosition}
            isEditingBPM={this.state.isEditingBPM}
            editedBPM={this.state.editedBPM}
            handleBPMChange={this.handleBPMChange}
            inputRef={this.inputRef}
            enterBPMEditingMode={this.enterBPMEditingMode}
            bpm={this.props.bpm}
        />
    }
}

const mapStateToProps = state => ({
    isPlaying: state.playerInfo.isPlaying,
    isMuted: state.playerInfo.isMuted,
    bpm: state.playerInfo.bpm,
    volume: state.playerInfo.volume
});

export default connect(
    mapStateToProps,
    {
        playTrack: ActionCreators.playTrack,
        stopTrack: ActionCreators.stopTrack,
        setBPM: ActionCreators.setBPM
    }
)(TransportContainer);