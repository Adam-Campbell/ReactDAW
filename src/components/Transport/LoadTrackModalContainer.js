import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';  
import * as ActionCreators from '../../actions';
import LoadTrackModal from './LoadTrackModal';

export class LoadTrackModalContainer extends Component {

    constructor(props) {
        super(props);
        this.root = document.getElementById('modal-root');
        this.el = document.createElement('div');
    }

    componentDidMount() {
        this.root.appendChild(this.el);
    }

    componentWillUnmount() {
        this.root.removeChild(this.el);
    }

    getSavedTracks() {
        const savedTracks = [];
        const len = window.localStorage.length;
        for (let i = 0; i < len; i++) {
            const key = window.localStorage.key(i);
            if (key.includes('[track_name]')) {
                const trimmedName = key.replace(/\[track_name\] /, '');
                savedTracks.push(trimmedName);
            }
        }
        return savedTracks;
    }

    handleLoad = (trackName) => {
        const namespacedTrackName = `[track_name] ${trackName}`;
        this.props.loadState(namespacedTrackName);
        this.props.handleClose();
    }

    render() {
        const savedTracks = this.getSavedTracks();
        return ReactDOM.createPortal(
            <LoadTrackModal 
                handleClose={this.props.handleClose}
                savedTracks={savedTracks}
                handleLoad={this.handleLoad}
            />,
            this.el
        );
    }
}

LoadTrackModalContainer.propTypes = {
    handleClose: PropTypes.func.isRequired
};

export default connect(
    undefined,
    {
        loadState: ActionCreators.loadState
    } 
)(LoadTrackModalContainer);
