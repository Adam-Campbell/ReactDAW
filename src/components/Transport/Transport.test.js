import React from 'react';
import { shallow } from 'enzyme';
import { TransportContainer } from './';
import Transport from './Transport';
import Tone from 'tone';

jest.mock('tone');
Tone.Transport = { position: '0:0:0' };

test('container component renders correctly', () => {
    const component = shallow(
        <TransportContainer 
            isPlaying={false}
            isPaused={false}
            isMuted={false}
            bpm={120}
            volume={0}
            playTrack={jest.fn()}
            stopTrack={jest.fn()}
            pauseTrack={jest.fn()}
            setBPM={jest.fn()}
            saveState={jest.fn()}
            loadState={jest.fn()}
            openWindow={jest.fn()}
        />
    );
    expect(component).toMatchSnapshot();
    expect(component.find(Transport)).toHaveLength(1);
});

test('container component dispatches playTrack and stopTrack actions correctly', () => {
    const mockedPlayTrack = jest.fn();
    const mockedStopTrack = jest.fn();
    const mockedPauseTrack = jest.fn();
    const component = shallow(
        <TransportContainer 
            isPlaying={false}
            isPaused={false}
            isMuted={false}
            bpm={120}
            volume={0}
            playTrack={mockedPlayTrack}
            stopTrack={mockedStopTrack}
            pauseTrack={mockedPauseTrack}
            setBPM={jest.fn()}
            saveState={jest.fn()}
            loadState={jest.fn()}
            openWindow={jest.fn()}
        />
    );
    expect(mockedPlayTrack).not.toHaveBeenCalled();
    component.first().shallow().find('button').at(0).simulate('click');
    expect(mockedPlayTrack).toHaveBeenCalledTimes(1);
    expect(mockedPauseTrack).not.toHaveBeenCalled();
    component.first().shallow().find('button').at(1).simulate('click');
    expect(mockedPauseTrack).toHaveBeenCalledTimes(1);
    expect(mockedStopTrack).not.toHaveBeenCalled();
    component.first().shallow().find('button').at(2).simulate('click');
    expect(mockedStopTrack).toHaveBeenCalledTimes(1);
});

test('presentational component renders correctly', () => {
    const component = shallow(
        <Transport 
            handleTransportBarClick={jest.fn()}
            playTrack={jest.fn()}
            stopTrack={jest.fn()}
            pauseTrack={jest.fn()}
            handleSkipToStart={jest.fn()}
            transportPosition={'1:2:3'}
            isEditingBPM={false}
            editedBPM={120}
            handleBPMChange={jest.fn()}
            inputRef={React.createRef()}
            enterBPMEditingMode={jest.fn()}
            bpm={120}
            handleLoadState={jest.fn()}
            handleSaveState={jest.fn()}
            enterSavingState={jest.fn()}
            exitSavingState={jest.fn()}
            enterLoadingState={jest.fn()}
            exitLoadingState={jest.fn()}
            isLoading={false}
            isSaving={false}
            handleOpenMixer={jest.fn()}
        />
    );
    expect(component).toMatchSnapshot();
});

