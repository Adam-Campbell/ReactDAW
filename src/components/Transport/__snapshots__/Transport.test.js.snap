// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container component renders correctly 1`] = `
<Transport
  bpm={120}
  editedBPM={120}
  enterBPMEditingMode={[Function]}
  handleBPMChange={[Function]}
  handleTransportBarClick={[Function]}
  inputRef={
    Object {
      "current": null,
    }
  }
  isEditingBPM={false}
  playTrack={[MockFunction]}
  stopTrack={[MockFunction]}
  transportPosition="0:0:0"
/>
`;

exports[`presentational component renders correctly 1`] = `
<div
  className="transport__container"
  onClick={[MockFunction]}
>
  <button
    className="button pink"
    onClick={[MockFunction]}
  >
    Play
  </button>
  <button
    className="button pink"
    onClick={[MockFunction]}
  >
    Stop
  </button>
  <span
    className="transport__track-position"
  >
    1:2:3
  </span>
  <div
    className="transport__bpm-container"
  >
    <span
      className="transport__bpm-label"
    >
      BPM:
    </span>
    <p
      className="transport__bpm-text"
      onClick={[MockFunction]}
    >
      120
    </p>
  </div>
</div>
`;
